### PR TITLE
UCT/IB: Use KSM for atomic_mr

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**
If atomic MW should cover MR larger than 2G - use KSM if supported.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6790/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6791/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6791/label=hpc-test-node2,worker=3/artifact/cov_build_6791/output/errors/index.html

File: src/uct/ib/base/ib_md.c [hunk 1]
```diff
@@ -347,12 +347,14 @@ static ucs_status_t uct_ib_md_post_umr(uct_ib_md_t *md, struct ibv_mr *mr,
...
 {
 #if HAVE_EXP_UMR
-    struct ibv_exp_mem_region mem_reg;
+    struct ibv_exp_mem_region *mem_reg = NULL;
     struct ibv_exp_send_wr wr, *bad_wr;
     struct ibv_exp_create_mr_in mrin;
     ucs_status_t status;
     struct ibv_mr *umr;
     struct ibv_wc wc;
+    int i = 0, list_size = 1;
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: better to initialize the vars right before use

File: src/uct/ib/base/ib_md.c [hunk 2]
```diff
@@ -365,14 +367,38 @@ static ucs_status_t uct_ib_md_post_umr(uct_ib_md_t *md, struct ibv_mr *mr,
...
+        IBV_EXP_DEVICE_ATTR_UMR_FIXED_SIZE_CAPS) {
+        reg_length = md->dev.dev_attr.umr_fixed_size_caps.max_entity_size;
+        list_size = ucs_div_round_up(mr->length, reg_length);
+    } else {
+        if (mr->length >= reg_length) {
+            status = UCS_ERR_UNSUPPORTED;
+            goto err;
+        }
+    }
+    mrin.attr.create_flags        = list_size > 1 ? IBV_EXP_MR_FIXED_BUFFER_SIZE
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: pls add ( ) around condition

File: src/uct/ib/base/ib_md.c [hunk 3]
```diff
@@ -365,14 +367,38 @@ static ucs_status_t uct_ib_md_post_umr(uct_ib_md_t *md, struct ibv_mr *mr,
...
     mrin.attr.create_flags        = IBV_EXP_MR_INDIRECT_KLMS;
+#endif
     mrin.attr.exp_access_flags    = UCT_IB_MEM_ACCESS_FLAGS;
-    mrin.attr.max_klm_list_size   = 1;
+    mrin.attr.max_klm_list_size   = list_size;
 #else
+    if (mr->length >= reg_length) {
+        status = UCS_ERR_UNSUPPORTED;
+        goto err;
+    }
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: pls add space line

File: src/uct/ib/base/ib_md.c [hunk 4]
```diff
@@ -365,14 +367,38 @@ static ucs_status_t uct_ib_md_post_umr(uct_ib_md_t *md, struct ibv_mr *mr,
     mrin.pd                       = md->pd;
 
 #ifdef HAVE_EXP_UMR_NEW_API
+#ifdef HAVE_EXP_UMR_KSM
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: for cleanup, let's remove #ifdef HAVE_EXP_UMR_NEW_API, and always assume it's the new API (and update ib.m4 accordingly)

File: src/uct/ib/base/ib_md.c [hunk 5]
```diff
@@ -365,14 +367,38 @@ static ucs_status_t uct_ib_md_post_umr(uct_ib_md_t *md, struct ibv_mr *mr,
...
+    if (md->dev.dev_attr.exp_device_cap_flags &
+        IBV_EXP_DEVICE_ATTR_UMR_FIXED_SIZE_CAPS) {
+        reg_length = md->dev.dev_attr.umr_fixed_size_caps.max_entity_size;
+        list_size = ucs_div_round_up(mr->length, reg_length);
+    } else {
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: can we do "} else if (..) {" ?

File: src/uct/ib/base/ib_md.c (hunk 2 above)
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: pls add space line before

File: src/uct/ib/base/ib_md.c [hunk 6]
```diff
@@ -365,14 +367,38 @@ static ucs_status_t uct_ib_md_post_umr(uct_ib_md_t *md, struct ibv_mr *mr,
     mrin.pd                       = md->pd;
 
 #ifdef HAVE_EXP_UMR_NEW_API
+#ifdef HAVE_EXP_UMR_KSM
+    if (md->dev.dev_attr.exp_device_cap_flags &
+        IBV_EXP_DEVICE_ATTR_UMR_FIXED_SIZE_CAPS) {
+        reg_length = md->dev.dev_attr.umr_fixed_size_caps.max_entity_size;
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: what if buffer size is smaller?

File: src/uct/ib/base/ib_md.c [hunk 7]
```diff
@@ -383,31 +409,35 @@ static ucs_status_t uct_ib_md_post_umr(uct_ib_md_t *md, struct ibv_mr *mr,
...
 
-    mem_reg.base_addr                              = (uintptr_t) mr->addr;
-    mem_reg.length                                 = mr->length;
+    mem_reg[0].base_addr                           = (uintptr_t) mr->addr;
 
 #ifdef HAVE_EXP_UMR_NEW_API
-    mem_reg.mr                                     = mr;
+    mem_reg[0].mr                                  = mr;
+    for (i = 1; i < list_size; i++) {
+        mem_reg[i-1].length                        = reg_length;
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: pls add spaces around "-"

File: src/uct/ib/base/ib_md.c [hunk 8]
```diff
@@ -347,12 +347,14 @@ static ucs_status_t uct_ib_md_post_umr(uct_ib_md_t *md, struct ibv_mr *mr,
...
 #if HAVE_EXP_UMR
-    struct ibv_exp_mem_region mem_reg;
+    struct ibv_exp_mem_region *mem_reg = NULL;
     struct ibv_exp_send_wr wr, *bad_wr;
     struct ibv_exp_create_mr_in mrin;
     ucs_status_t status;
     struct ibv_mr *umr;
     struct ibv_wc wc;
+    int i = 0, list_size = 1;
+    size_t reg_length = 0x80000000;
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: is it because we don't have max KLM size?
if so let's define as constant is ib_device.h

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6827/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@Artemy-Mellanox pls also add a test which allocates atomic on >2GB offset

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6836/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6840/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_md.c [hunk 9]
```diff
@@ -347,71 +343,92 @@ static ucs_status_t uct_ib_md_post_umr(uct_ib_md_t *md, struct ibv_mr *mr,
...
+    wr.ext_op.umr.exp_access            = UCT_IB_MEM_ACCESS_FLAGS;
+
+    reg_length = UCT_IB_MD_MAX_MR_SIZE;
+#ifdef HAVE_EXP_UMR_KSM
+    if ((md->dev.dev_attr.comp_mask & IBV_EXP_DEVICE_ATTR_COMP_MASK_2) &&
+        (md->dev.dev_attr.comp_mask_2 & IBV_EXP_DEVICE_ATTR_UMR_FIXED_SIZE_CAPS) &&
+        (md->dev.dev_attr.exp_device_cap_flags & IBV_EXP_DEVICE_UMR_FIXED_SIZE))
+    {
+        reg_length = md->dev.dev_attr.umr_fixed_size_caps.max_entity_size;
+        list_size = ucs_div_round_up(mr->length, reg_length);
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: pls indent on column

File: src/uct/ib/base/ib_md.c [hunk 10]
```diff
@@ -445,11 +462,13 @@ static ucs_status_t uct_ib_md_post_umr(uct_ib_md_t *md, struct ibv_mr *mr,
               umr->lkey, umr->rkey);
     *umr_p = umr;
 
+    ucs_free(mem_reg);
     return UCS_OK;
 
 err_free_umr:
     UCS_PROFILE_CALL(ibv_dereg_mr, umr);
 err:
+    ucs_free(mem_reg);
```
[the remark this answers is not in the document]
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_md.c]: Is ucs_free(NULL) forbidden?

File: test/gtest/uct/ib/test_ib_md.cc [hunk 11]
```diff
@@ -113,4 +116,28 @@ UCS_TEST_P(test_ib_md, ib_md_umr_direct, "REG_METHODS=direct") {
...
+    ASSERT_TRUE(rkey_buffer != NULL);
+
+#ifdef HAVE_EXP_UMR_KSM
+    has_ksm = ucs_derived_of(pd(), uct_ib_md_t)->dev.dev_attr.exp_device_cap_flags &
+        IBV_EXP_DEVICE_UMR_FIXED_SIZE;
+#else
+    has_ksm = false;
+#endif
+
+    ib_md_umr_check(rkey_buffer, has_ksm, 0x80001000);
```
[the remark this answers is not in the document]
Comment (Artemy-Mellanox) [reply test/gtest/uct/ib/test_ib_md.cc]: this is the value which in absence of `HAVE_EXP_UMR_KSM` should cause disabling of atomic_mr

File: src/uct/ib/base/ib_md.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (brminich) [reply src/uct/ib/base/ib_md.c]: AFAIR, yes, with some compilers/options.
@hoopoepg, ping

[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: actually POSIX free(NULL) is allowed, but ucs_free may be expanded to custom/debug implementation which may crash on NULL pointer

[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/base/ib_md.c]: ucs_free(NULL) is allowed

File: test/gtest/uct/ib/test_ib_md.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (brminich) [reply test/gtest/uct/ib/test_ib_md.cc]: so this value should be bigger than `UCT_IB_MD_MAX_MR_SIZE`, right? Maybe change it to `UCT_IB_MD_MAX_MR_SIZE + X` then?

File: src/uct/ib/base/ib_md.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (brminich) [reply src/uct/ib/base/ib_md.c]: ok

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6974/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:bgate:retest

File: src/uct/ib/base/ib_verbs.h [hunk 12]
```diff
@@ -60,7 +60,14 @@
 
 #else
 #  define IBV_SHARED_MR_ACCESS_FLAGS(_shared_mr)    ((_shared_mr)->access)
+#ifndef HAVE_EXP_UMR_KSM
```
Comment (yosefe) [inline src/uct/ib/base/ib_verbs.h]: minor: i think it's better to check for IBV_EXP_DEVICE_ATTR_RESERVED_2 explicitly

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6987/ for details (Mellanox internal link).

**Pull request merged**